tests/pfquickcheck/pflang.lua

```
@@ -0,0 +1,219 @@
+   end
+end
+
+local function packetAccessLocation()
+   local r1, r2 = math.random(), math.random()
+   local base
+   -- Usually generate small accesses - more likely to be in range
+   if r1 < 0.9 then
+      base = uint8()
+   else
+      base = uint16()
+   end
+   if r2 < 0.5 then
+      return base
+   else
+      -- tcpdump only allows the following 3 offsets
+      local offset = choose({1,2,4})
+      return base .. ':' .. offset
+   end
+end
+
+local function PacketAccess()
+   local proto = ProtocolName()[1]
+   -- Avoid packet access on protocols where libpcap doesn't allow it
+   -- libpcap does not allow 'ah' and 'esp' packet access; not a pflua bug
+   -- libpcap does not allow icmp6[x]:
+   -- "IPv6 upper-layer protocol is not supported by proto[x]"
+   local skip_protos = set('ah', 'esp', 'icmp6')
+   while skip_protos[proto] do
+      proto = ProtocolName()[1]
+   end
+   local access = packetAccessLocation()
+   local comparison = table.concat({comparisonOp(), uint8()}, ' ')
+   local pkt_access = table.concat({proto, '[', access, '] '})
+   return {pkt_access, comparisonOp(), uint8()}
+end
+
+local function PflangClause()
+   return choose({ProtocolName, Port, PortRange, ProtocolWithPort,
+                  Host, Net, Math, LenWithMath, PacketAccess})()
+end
+
+-- Add logical operators (or/not)
+-- Do not test 'and', because it fails fast on the BPF pipeline for some
+-- cases that match no packets. (Bug 130)
+function PflangLogical()
+   local r = math.random()
+   local pclause = PflangClause()
+   local pclause2 = PflangClause()
+
+   if r < 0.9 then
+      local logicOp = 'or'
+
+      table.insert(pclause, logicOp)
+      for _,v in ipairs(pclause2) do table.insert(pclause, v) end
+      return pclause
+   else
+      table.insert(pclause, 1, "not")
+      return pclause
+   end
+end
+
+function Pflang()
+   local r = math.random()
+   if r < 0.001 then return Empty() end
+   -- Work around bugs 131 and 132, as well as 166/171
```

Yep. It's now actually waiting for me to hack around libpcap's fail fast mechanism, because things like "not 1 = 1" can be generated, and reject all packets.